### DISCUSSION TITLE
Transferring big data chunks between device and phone

### DISCUSSION THREAD

**Discussion by laurislopata** [kenjdavidson/react-native-bluetooth-classic#183] [Q&A] [answered]
Hello,

I have a general question to ask about an implementation of a protocol my company will need. On a high level, we're implementing an update system for a device that will have a Bluetooth connection with the phone.

To update the firmware of the device, we will send a HTTP request to a server and the server will respond with the new firmware of the device. Then to minimise the complexity we would want to simply parse the response and immediately transmit it as a byte array over a Bluetooth connection to the device. The size of this updated firmware would be around 40MB. 

My question is if this type of custom data transmission could be done just by using this library or would we need to implement it using native language (kotlin and swift)?

Thank you in advance for the replies and have a nice upcoming end of the week to anyone who's reading this question :)

**Comment by kenjdavidson** [comment]
Assuming that you have no issues with:
- Downloading the firmware file with React Native
- Accepting the byte array and installing the firmware once complete on the device

It should be possible to use this library to perform the:
- Connecting to device
- Sending byte array data you've already downloaded
- Closing the connection

Logically you should be able to use the library.  

There is a super simple `ByteArrayDeviceConnectionImpl` that is available, although I don't know how much it's been tested.  It's also only available on Android, I have not implemented it on IOS at this point.  At worst case, you'd need to write your own `DeviceConnection` class for both Android or IOS, but I've attempted to write the library in a way that makes this pretty straight forward.

I'll be interested to hear how it works out, but from what you've said, there should be nothing stopping you from using the library.

**Reply by laurislopata** [reply]
Thank you very much, just out of curiosity. Are there any plans in the future to implement the same functionality as `ByteArrayDeviceConnectionImpl` for iOS?

**Comment by laurislopata** [comment]
Another question I have if you could help me a bit. For the iOS part on a high level what would have to be implemented to support the same implementation that Android currently has?  The documentation of Apple is pretty average, maybe you know some build libraries for using Bluetooth with swift, or know some good resource to look at? Thank in advance for your answers, you're relly helpful :)

**Comment by kenjdavidson** [comment, marked as answer]
Honestly the IOS side of things is really light, it took me a bunch of reading (mainly in Objective C) in order to convert things to Swift and get working.  There are two things that need to be done:

1. You'll need to figure out (and walk me through) the Auto Loading overrides, as I don't think that's programmed into the IOS side at this point.
2. You'll need to write the ByteArrayDeviceConnectionImpl on the IOS side, that manages byte arrays instead of String data.

I'll definitely take a look/accept pull requests if you get it working, but I don't have the time to look into it.

**Reply by laurislopata** [reply]
Thank you for your answer, I'll keep you posted if we manage to make this work

**Reply by kenjdavidson** [reply]
Here is the autolinking document that I found helpful
https://github.com/react-native-community/cli/blob/master/docs/autolinking.md

Android allows you to easily:
- turn off autolinking
- modify the Android app to include your own custom `Package` class

On IOS I haven't figured out how to do this, but maybe you can.  Essentially you want to be bale t

```
        self.connectionFactories = Dictionary()
        self.connectionFactories["delimited"] = DelimitedStringDeviceConnectionFactory()
        // put your version here 
        self.connectionFactories["binary"] = ByteArrayDeviceDeviceConnectionFactory()
```

I remember reading earlier, and there is a way in your application to request (on the native app) to get a list of modules loaded, and then modify them.  If this is still available, that's how you would be able to add to the current `connectionFactories`.

**Reply by kenjdavidson** [reply]
If you get it working, it's already put into the Android side by default, so we can definitely put the same implementation into IOS by default, so that we don't need to disable auto linking for binary/bytearray connections.
